lib/Subscriber.js

```diff
@@ -0,0 +1,592 @@
+    }
+    else 
+    {
+      this._sendSubsequentSubscribe(body, this._headers);
+    }
+  }
+
+  /** 
+   * terminate. 
+   * Send un-subscribe or fetch-subscribe (with Expires: 0).
+   * @param {string} body - un-subscribe request body
+   */
+  terminate(body = null) 
+  {
+    logger.debug('terminate()');
+
+    // Prevent duplication un-subscribe sending.
+    if (this._send_unsubscribe)
+    {
+      logger.warn('unsubscribe has already been sent');
+
+      return;
+    }
+    this._send_unsubscribe = true;
+
+    // Set header Expires: 0.
+    const headers = this._headers.map((s) => 
+    { 
+      return s.startsWith('Expires') ? 'Expires: 0' : s; 
+    });
+
+    if (this._state === C.STATE_INIT)
+    {
+      // fetch-subscribe - initial subscribe with Expires: 0.
+      this._sendInitialSubscribe(body, headers);
+    }
+    else
+    {  
+      this._sendSubsequentSubscribe(body, headers);
+    }
+
+    // Waiting for the final notify for a while.
+    const final_notify_timeout = 30000;
+
+    this._unsubscribe_timeout_timer = setTimeout(() => 
+    {
+      this._dialogTerminated(C.UNSUBSCRIBE_TIMEOUT);
+    }, final_notify_timeout);
+  }
+
+  /**
+   * Get dialog state.
+   */
+  get state()
+  {
+    return this._state;
+  }
+  
+  /**
+   * Get dialog id.
+   */
+  get id()
+  {
+    return this._dialog ? this._dialog.id : null;
+  }
+
+  /**
+   * Private API.
+   */
+  _sendInitialSubscribe(body, headers)
+  {
+    if (body) 
+    {
+      if (!this._content_type)
+      { 
+        throw new TypeError('content_type is undefined');
+      }
+
+      headers = headers.slice();
+      headers.push(`Content-Type: ${this._content_type}`);
+    }
+
+    this._state = C.STATE_NOTIFY_WAIT;
+
+    const request = new SIPMessage.OutgoingRequest(JsSIP_C.SUBSCRIBE, 
+      this._ua.normalizeTarget(this._target), this._ua, this._params, headers, body);
+
+    const request_sender = new RequestSender(this._ua, request, {
+      onRequestTimeout : () =>
+      {
+        this.onRequestTimeout();
+      },
+      onTransportError : () =>
+      {
+        this.onTransportError();
+      },
+      onReceiveResponse : (response) =>
+      {
+        this._receiveSubscribeResponse(response);
+      }
+    });
+
+    request_sender.send();
+  }
+
+  _receiveSubscribeResponse(response) 
+  {
+    if (response.status_code >= 200 && response.status_code < 300) 
+    {
+      // Create dialog
+      if (this._dialog === null) 
+      {
+        const dialog = new Dialog(this, response, 'UAC');
+
+        if (dialog.error)
+        {
+          // OK response without Contact 
+          logger.warn(dialog.error);
+          this._dialogTerminated(C.SUBSCRIBE_BAD_OK_RESPONSE);
+
+          return;
+        }
+        
+        this._dialog = dialog;  
+
+        logger.debug('emit "dialogCreated"');
+        this.emit('dialogCreated');
+
+        // Subsequent subscribes saved in the queue until dialog created.
```

Comment:
Instead of enqueuing. We should not send any subsequent subscribe until the initial was accepted. Hence the logic is placed in the application:

subscribe.on('accepted', () => { send another subscribe})

Reply:
I'm agree - it is good if the user writes code in this style.
So better have such event.

However, it is good also to have subsequence enqueue/dequeue code in subscriber.

For example, this will allow to call the subscribe.terminate() at any time (even before the event 'accepted')

E.g. user send initial subscribe, and after this suddenly want terminate subscription,
in the case he call subscriber.terminate() and don't think about events and state machine...